lib.go

```diff
@@ -51,9 +51,15 @@ func (vm *VM) Cleanup() {
 // This function stores the code for that contract only once, but it can
 // be instantiated with custom inputs in the future.
 //
-// TODO: return gas cost? Add gas limit??? there is no metering here...
-func (vm *VM) StoreCode(code WasmCode) (Checksum, error) {
-	return api.StoreCode(vm.cache, code)
+// Returns both the checksum, as well as the gas cost of compilation (in CosmWasm Gas) or an error.
+func (vm *VM) StoreCode(code WasmCode, gasLimit uint64) (Checksum, uint64, error) {
+	gasCost := compileCosts(code)
+	if gasLimit < gasCost {
+		return nil, gasCost, fmt.Errorf("insufficient gas to store contract (%d bytes)", len(code))
```

Comment:
Can we use a typed error here, like `types.OutOfGasError{}`? This at least is what `func errorWithMessage` is doing when an out of gas happens in the VM.

Reply:
Yes, we can do that. I was using the deserialization as a template here.